README.txt
======================
silva.security.logging
======================

Introduction
============

This extensions log in details user actions in Silva. To do so, you
need to add a *Silva Security Logging Service* in ZMI, in a site
(either your Silva Root or a local site).

You can configure the logging output. By default it will be logged in
the Zope logs, but you can log to an SQL database as well.

In order to do this, you need to configure a SQL connection in Zope to
your database, and in *Silva Security Logging Service* to select SQL
logging, and your database identifier. The connected database should
have a table called ``log`` created like this::

  create table log (
      username varchar(255),
      action varchar(255),
      time datetime,
      content varchar(512),
      content_intid bigint,
      info varchar(255))


.. note::

   This table is not automatically created for you.

.. note::

   You can name your table differently, and configure the name on the
   service.

The log storage is extensible, you can provide your own storage.


Code repository
===============

You can find the code for this extension in Mercurial:
https://hg.infrae.com/silva.security.logging/.


docs/HISTORY.txt
Changes
=======

1.3 (unreleased)
----------------

* ...

1.2.2 (2012/03/12)
------------------

* Remove debug warning.

1.2.1 (2012/03/01)
------------------

* Make sure that entries contains only strings, in order to create
  valid SQL queries.

1.2 (2012/02/23)
----------------

* Make sure the code works on both Silva 2.3 and 3.0.

* Disable logging during upgrade operations (for speed).

* Prevent breakage when storage is misconfigured.

* Add possibility to configure table name with SQL storage.


1.1 (2011/11/14)
----------------

* Update code for Silva 3.0.

* Add more events to by logged (there are more in 3.0).

1.0 (2010/12/20)
----------------

* Initial release.



src/silva/security/logging/tests/test_events.py
# -*- coding: utf-8 -*-
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

import unittest

import DateTime
from Products.Silva.tests.helpers import publish_content

from zope.component import queryUtility
from silva.security.logging.testing import FunctionalLayer
from silva.security.logging.interfaces import ISecurityLoggingService


class AssertLog(object):

    def __init__(self, test, expected):
        self.test = test
        self.expected = expected

    def __enter__(self):
        self.test.logger.purge()

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.test.assertEqual(self.expected, self.test.logger.records)
        self.test.logger.purge()


class SubscribersTestCase(unittest.TestCase):
    layer = FunctionalLayer

    def setUp(self):
        self.root = self.layer.get_application()
        self.layer.login('editor')

        factory = self.root.manage_addProduct['silva.security.logging']
        factory.manage_addSecurityLoggingService()

        service = queryUtility(ISecurityLoggingService)
        service.storage_name = 'Memory Logger (for testing)'
        self.logger = service.get_logger()

    def tearDown(self):
        self.logger.purge()

    def test_add_content(self):
        with AssertLog(
            self,
            [['editor', 'add', '/root/folder'],
             ['editor', 'modify the container', '/root'],
             ['editor', 'create', '/root/folder']]):
            factory = self.root.manage_addProduct['Silva']
            factory.manage_addFolder('folder', 'Folder')

    def test_copy_content(self):
        factory = self.root.manage_addProduct['Silva']
        factory.manage_addFolder('folder', 'Folder')

        with AssertLog(
            self,
            [['editor', 'copy', 'copy_of_folder', 'from /root/folder'],
             ['editor', 'add', '/root/copy_of_folder'],
             ['editor', 'modify the container', '/root']]):
            token = self.root.manage_copyObjects(['folder'])
            self.root.manage_pasteObjects(token)

    def test_rename_content(self):
        factory = self.root.manage_addProduct['Silva']
        factory.manage_addFolder('folder', 'Folder')

        with AssertLog(
            self,
            [['editor', 'move', '/root/news', 'from /root/folder to /root/news'],
             ['editor', 'modify the container', '/root']]):
            self.root.manage_renameObjects(['folder'], ['news'])

    def test_move_content(self):
        factory = self.root.manage_addProduct['Silva']
        factory.manage_addFolder('folder', 'Folder')
        factory.manage_addFolder('target', 'Target')

        with AssertLog(
            self,
            [['editor', 'move', '/root/target/folder', 'from /root/folder to /root/target/folder'],
             ['editor', 'modify the container', '/root'],
             ['editor', 'modify the container', '/root/target']]
):
            token = self.root.manage_cutObjects(['folder'])
            self.root.target.manage_pasteObjects(token)

    def test_delete_content(self):
        factory = self.root.manage_addProduct['Silva']
        factory.manage_addFolder('folder', 'Folder')

        with AssertLog(
            self,
            [['editor', 'delete', '/root/folder'],
             ['editor', 'modify the container', '/root']]):
            self.root.manage_delObjects(['folder'])

    def test_publication(self):
        factory = self.root.manage_addProduct['SilvaDocument']
        factory.manage_addDocument('document', 'Document')
        document = self.root.document

        with AssertLog(
            self,
            [['editor', 'publish', '/root/document/0']]):
            # First publication
            publish_content(document)

        with AssertLog(
            self,
            [['editor', 'add', '/root/document/1'],
             ['editor', 'modify the container', '/root/document']]):
            # Create copy
            document.create_copy()

        with AssertLog(
            self,
            [['editor', 'close', '/root/document/0'],
             ['editor', 'publish', '/root/document/1']]):
            # Publish copy (so close the published one first)
            publish_content(document)

    def test_request_approval(self):
        self.layer.login('author')
        factory = self.root.manage_addProduct['SilvaDocument']
        factory.manage_addDocument('document', 'Document')

        document = self.root.document
        document.set_unapproved_version_publication_datetime(
            DateTime.DateTime() - 10)

        with AssertLog(
            self,
            [['author', 'request approval', '/root/document/0']]):
            # Request approval
            document.request_version_approval('Ready')

        with AssertLog(
            self,
            [['author', 'cancel request approval', '/root/document/0']]):
            # Withdraw approval
            document.withdraw_version_approval('Not really ready')

        document.request_version_approval('Ready')
        self.layer.login('chiefeditor')

        with AssertLog(
            self,
            [['chiefeditor', 'cancel request approval', '/root/document/0']]):
            # Reject approval
            document.withdraw_version_approval('Not really ready')


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(SubscribersTestCase))
    return suite


src/silva/security/logging/tests/test_service.py
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

import unittest

from Products.Silva.testing import SilvaLayer

from zope.component import queryUtility
from zope.interface.verify import verifyObject
from silva.security.logging.interfaces import ISecurityLoggingService
import silva.security.logging


class ServiceTestCase(unittest.TestCase):

    layer = SilvaLayer(silva.security.logging)

    def setUp(self):
        self.root = self.layer.get_application()

    def test_add(self):
        """We add a service and check we can retrieve it.
        """
        self.failIf(hasattr(self.root, 'service_securitylogging'))
        service = queryUtility(ISecurityLoggingService)
        self.assertEqual(None, service)

        factory = self.root.manage_addProduct['silva.security.logging']
        factory.manage_addSecurityLoggingService()
        self.failUnless(hasattr(self.root, 'service_securitylogging'))

        service = queryUtility(ISecurityLoggingService)
        self.assertEqual(self.root.service_securitylogging, service)
        self.failUnless(verifyObject(ISecurityLoggingService, service))

        self.root.manage_delObjects(['service_securitylogging'])
        service = queryUtility(ISecurityLoggingService)
        self.assertEqual(None, service)


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(ServiceTestCase))
    return suite


src/silva/security/logging/entry.py
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

from AccessControl import getSecurityManager

from five import grok
from zope.intid.interfaces import IIntIds
from zope.keyreference.interfaces import NotYet
from zope.component import queryUtility

from silva.security.logging.interfaces import (
    ILoggingEvent, ISecurityLoggingService)


def get_username():
    """Return the username of the current user.
    """
    return getSecurityManager().getUser().getUserName()


def get_path(content):
    """Return the path in Zope for the content.
    """
    return '/'.join(content.getPhysicalPath())


def get_id(content):
    """Return an unique identifier for the content, in order to be
    able to find it again even if it moves.
    """
    service = queryUtility(IIntIds)
    if service is not None:
        try:
            return service.register(content)
        except NotYet:
            return 0
    return 0


class LoggingEvent(object):
    """Represent an event.
    """
    grok.implements(ILoggingEvent)

    def __init__(self, action, content, detail=None):
        self.action = action
        self.username = get_username()
        self.content_path = get_path(content)
        self.content_id = get_id(content)
        self.detail = detail
        self._service = None

    def get_service(self):
        if self._service is None:
            self._service = queryUtility(ISecurityLoggingService)
        return self._service

    def log(self):
        service = self.get_service()
        if service is not None:
            logger = service.get_logger()
            if logger is not None:
                logger.log(self)

    def enable(self):
        service = self.get_service()
        if service is not None:
            service.enable_logging = True

    def disable(self):
        service = self.get_service()
        if service is not None:
            service.enable_logging = False


src/silva/security/logging/service.py
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

from five import grok
from persistent.dict import PersistentDict
from zeam.form import silva as silvaforms
from zeam.form.base.datamanager import DictDataManager
from zope.component import getUtility

from silva.core import conf as silvaconf
from silva.core.views import views as silvaviews
from silva.core.services.base import SilvaService
from silva.security.logging.interfaces import (
    ISecurityLoggingService, ILoggingStorage, ISecurityLoggingConfiguration)


class SecurityLoggingService(SilvaService):
    """Service to logging security events.
    """
    meta_type = "Silva Security Logging Service"
    default_service_identifier = 'service_securitylogging'
    grok.implements(ISecurityLoggingService)
    silvaconf.icon('service.png')

    manage_options = (
        {'label': 'Access log', 'action': 'manage_main'},
        {'label': 'Configuration', 'action': 'manage_config'},
        ) + SilvaService.manage_options

    def __init__(self, identifier):
        super(SecurityLoggingService, self).__init__(identifier)
        self.storage_name = "Python Logger"
        self.storage_conf = PersistentDict()

    def get_storage(self):
        return getUtility(ILoggingStorage, name=self.storage_name)

    def get_logger(self):
        return self.get_storage().configure(self)


class ViewLog(silvaviews.ZMIView):
    """View logs.
    """
    grok.name('manage_main')

    def update(self):
        self.is_viewable = False


class Configuration(silvaforms.ZMIComposedForm):
    """Configure service.
    """
    grok.name('manage_config')

    label = u"Change configuration"
    description = u"Modify storage and storage options"



class SelectStorage(silvaforms.SubForm):
    silvaforms.view(Configuration)

    label = u"Select storage to use"
    fields = silvaforms.Fields(ISecurityLoggingConfiguration)
    actions = silvaforms.Actions(silvaforms.EditAction(u"Change"))
    ignoreContent = False


class StorageConfiguration(silvaforms.SubForm):
    silvaforms.view(Configuration)

    label = u"Configure storage"
    actions = silvaforms.Actions(silvaforms.EditAction(u"Change"))
    ignoreContent = False

    @property
    def fields(self):
        return self.context.get_storage().storage_conf

    def update(self):
        self.setContentData(DictDataManager(self.context.storage_conf))




src/silva/security/logging/storage.py
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

import logging

from five import grok
from zeam.form import silva as silvaforms

from silva.security.logging.interfaces import (
    ILoggingStorage, ISQLStorageConfiguration, ILogger)

logger = logging.getLogger("silva.security.logging")


class PythonLogger(grok.GlobalUtility):
    grok.implements(ILoggingStorage, ILogger)
    grok.provides(ILoggingStorage)
    grok.name('Python Logger')

    storage_conf = silvaforms.Fields()

    def log(self, event):
        msg = u'%r did %s on %s'
        info = [event.username, event.action, event.content_path]
        if event.detail:
            msg += u' (%s)'
            info.append(event.detail)
        logger.info(msg % tuple(info))

    def configure(self, context):
        return self


class ZopeSQLLogger(object):
    """Log in a Zope SQL connection.
    """
    grok.implements(ILogger)

    def __init__(self, connection, table='log'):
        self.connection = connection
        self.table = table

    def log(self, event):
        db = self.connection()
        query = """insert into %s
                        (username, action, time, content, content_intid, info)
                        values (%r, %r, now(), %r, %d, %r)""" % (
                self.table, event.username, event.action,
                event.content_path, event.content_id,
                event.detail or "")
        logger.warn(query)
        db.query(query)


class ZopeSQLStorage(grok.GlobalUtility):
    grok.implements(ILoggingStorage)
    grok.provides(ILoggingStorage)
    grok.name('Zope SQL')

    storage_conf = silvaforms.Fields(ISQLStorageConfiguration)

    def configure(self, context):
        if 'sql_connection_id' not in context.storage_conf:
            logger.error('SQL logger not properly configured')
            return None
        connection_id = context.storage_conf['sql_connection_id']
        table = str(context.storage_conf.get('sql_table', 'log'))
        try:
            connection = getattr(context, connection_id)
        except AttributeError:
            logger.error('SQL connection %s no longer exists' % connection_id)
            return None
        return ZopeSQLLogger(connection, table)


class MemoryLogger(grok.GlobalUtility):
    grok.implements(ILoggingStorage, ILogger)
    grok.provides(ILoggingStorage)
    grok.name('Memory Logger (for testing)')

    storage_conf = silvaforms.Fields()
    records = []                # Records will be stored on the class

    def log(self, event):
        info = [event.username, event.action, event.content_path]
        if event.detail:
            info.append(event.detail)
        self.records.append(info)

    def purge(self):
        del self.records[:]

    def configure(self, context):
        return self


src/silva/security/logging/interfaces.py
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

from five import grok
from zope import schema
from zope.component import getUtilitiesFor
from zope.interface import Interface, Attribute
from zope.schema.vocabulary import SimpleVocabulary, SimpleTerm
from zope.schema.interfaces import IContextSourceBinder

from Products.ZSQLMethods.SQL import SQLConnectionIDs
from Acquisition import aq_inner

from silva.core.interfaces import ISilvaLocalService


class ISecurityEvent(Interface):
    """A security event.
    """


class ILoggingStorage(Interface):
    """Describe a logging storage.
    """
    storage_conf = Attribute(u"Configuration fields for this storage")

    def configure(context):
        """Return a logger for the given context.
        """


class ILogger(Interface):
    """A logger able to log an event.
    """

    def log(event):
        """Log the given event.
        """


@grok.provider(IContextSourceBinder)
def available_storages(context):
    return SimpleVocabulary(
        map(lambda (name, storage): SimpleTerm(value=name, title=name),
            getUtilitiesFor(ILoggingStorage)))


@grok.provider(IContextSourceBinder)
def available_sql_connections(context):
    return SimpleVocabulary(
        map(lambda (title, name): SimpleTerm(value=name, title=title),
            SQLConnectionIDs(aq_inner(context))))


class ISecurityLoggingConfiguration(Interface):
    """Configuration for the Security Logging service.
    """
    storage_name = schema.Choice(
        title=u"Log storage",
        source=available_storages)


class ISQLStorageConfiguration(Interface):
    """Configuration settings for the SQL storage
    """
    sql_connection_id = schema.Choice(
        title=u"SQL Connection",
        source=available_sql_connections)
    sql_table = schema.TextLine(
        title=u"SQL Table to use",
        default=u"log",
        required=True)


class ISecurityLoggingService(
    ISilvaLocalService, ISecurityLoggingConfiguration):
    """Local service to log security event.
    """
    storage_conf = Attribute(u"Configuration for the storage")

    def get_storage():
        """Return storage used to log events.
        """

    def get_logger():
        """Return storage's logger to log events.
        """


src/silva/security/logging/subscribers.py
# Copyright (c) 2010 Infrae. All rights reserved.
# See also LICENSE.txt
# $Id$

from five import grok
from zope.container.interfaces import (
    IObjectMovedEvent, IObjectAddedEvent, IObjectRemovedEvent,
    IContainerModifiedEvent)
from zope.lifecycleevent.interfaces import (
    IObjectCreatedEvent, IObjectCopiedEvent, IObjectModifiedEvent)

from OFS.interfaces import IObjectWillBeRemovedEvent

from silva.security.logging.entry import SecurityEvent, get_path
from silva.core.interfaces import events
from silva.core.interfaces import ISilvaObject, IVersion


@grok.subscribe(ISilvaObject, IObjectCreatedEvent)
def log_create(content, event):
    if IObjectCopiedEvent.providedBy(event):
        detail = 'from %s' % get_path(event.original)
        SecurityEvent('copy', content, detail).log()
    else:
        SecurityEvent('create', content).log()

@grok.subscribe(IVersion, IObjectWillBeRemovedEvent)
@grok.subscribe(ISilvaObject, IObjectWillBeRemovedEvent)
def log_delete(content, event):
    if content is event.object:
        SecurityEvent('delete', content).log()

@grok.subscribe(IVersion, IObjectAddedEvent)
@grok.subscribe(ISilvaObject, IObjectAddedEvent)
def log_add(content, event):
    if content is event.object:
        SecurityEvent('add', content).log()

@grok.subscribe(IVersion, IObjectModifiedEvent)
@grok.subscribe(ISilvaObject, IObjectModifiedEvent)
def log_modify(content, event):
    if not events.IPublishingEvent.providedBy(event):
        if IContainerModifiedEvent.providedBy(event):
            SecurityEvent('modify the container', content).log()
        else:
            SecurityEvent('modify', content).log()

@grok.subscribe(ISilvaObject, IObjectMovedEvent)
def log_move(content, event):
    if content is event.object:
        if not (IObjectAddedEvent.providedBy(event) or
                IObjectRemovedEvent.providedBy(event)):
            detail = 'from %s/%s to %s/%s' % (
                get_path(event.oldParent), event.oldName,
                get_path(event.newParent), event.newName)
            SecurityEvent('move', content, detail).log()

if hasattr(events, 'IContentOrderChangedEvent'):
    # Only available in Silva 3.0
    @grok.subscribe(ISilvaObject, events.IContentOrderChangedEvent)
    def log_order_changed(content, event):
        detail = 'from %s to %s' % (event.old_position, event.new_position)
        SecurityEvent('order changed', content, detail).log()


# Security changes

@grok.subscribe(ISilvaObject, events.ISecurityRestrictionModifiedEvent)
def log_security_restriction_set(content, event):
    if event.role:
        detail = 'with role %s' % event.role
        SecurityEvent('set a access restriction to', content, detail).log()
    else:
        SecurityEvent('remove the access restriction to', content).log()

@grok.subscribe(ISilvaObject, events.ISecurityRoleAddedEvent)
def log_security_add_role(content, event):
    detail = 'for %r roles %s' % (event.username, ', '.join(event.roles))
    SecurityEvent('changed roles on', content, detail).log()

@grok.subscribe(ISilvaObject, events.ISecurityRoleRemovedEvent)
def log_security_remove_role(content, event):
    detail = 'for %r' % event.username
    if event.roles:
        detail += ' roles %s' % ', '.join(event.roles)
    SecurityEvent('removed roles on', content, detail).log()


# Publication events

@grok.subscribe(IVersion, events.IContentRequestApprovalEvent)
def log_publication_request_approval(content, event):
    SecurityEvent('request approval', content).log()

if hasattr(events, 'IContentApprovalRequestWithdrawnEvent'):
    # Silva 3.0
    @grok.subscribe(IVersion, events.IContentApprovalRequestWithdrawnEvent)
    def log_publication_approval_request_withdrawn(content, event):
        SecurityEvent('cancel request approval', content).log()
else:
    # Silva 2.3+
    @grok.subscribe(IVersion, events.IContentApprovalRequestCanceledEvent)
    def log_publication_approval_request_cancel(content, event):
        SecurityEvent('cancel request approval', content).log()

@grok.subscribe(IVersion, events.IContentApprovalRequestRefusedEvent)
def log_publication_approval_request_refused(content, event):
    SecurityEvent('reject request approval', content).log()

@grok.subscribe(IVersion, events.IContentApprovedEvent)
def log_publication_approved(content, event):
    SecurityEvent('approve', content).log()

@grok.subscribe(IVersion, events.IContentUnApprovedEvent)
def log_publication_unapproved(content, event):
    SecurityEvent('revoke', content).log()

@grok.subscribe(IVersion, events.IContentPublishedEvent)
def log_publication_published(content, event):
    SecurityEvent('publish', content).log()

@grok.subscribe(IVersion, events.IContentClosedEvent)
def log_publication_closed(content, event):
    SecurityEvent('close', content).log()
